### PR TITLE
Fix upgrade test

### PR DESCRIPTION
**Pull request by amoskong**
This PR updated upgrade test to use scenario in official upgrade documentation, better support for scylla-enterprise, and introduced rollback subtest.

### CODE REVIEW

File: upgrade_test.py [hunk 1]
```diff
@@ -36,45 +37,57 @@ def upgrade_node(self, node):
...
             # node.remoter.run('sudo yum update -y --skip-broken', connect_timeout=900)
             node.remoter.run('sudo yum install python34-PyYAML -y')
             # replace the packages
             node.remoter.run('rpm -qa scylla\*')
+            # flush all memtables to SSTables
+            node.remoter.run('sudo nodetool drain')
             node.remoter.run('sudo nodetool snapshot')
+            node.remoter.run('sudo systemctl stop scylla-server.service')
             # update *development* packages
             node.remoter.run('sudo rpm -UvhR --oldpackage /tmp/scylla/*development*', ignore_status=True)
```
Comment (roydahan) [inline upgrade_test.py]: Do we still need this?
if not, remove

File: upgrade_test.py [hunk 2]
```diff
@@ -36,45 +37,57 @@ def upgrade_node(self, node):
...
             # and all the rest
             node.remoter.run('sudo rpm -URvh --replacefiles /tmp/scylla/* | true')
```
Comment (roydahan) [inline upgrade_test.py]: same

File: upgrade_test.py [hunk 3]
```diff
@@ -36,45 +37,57 @@ def upgrade_node(self, node):
...
             node.remoter.run('rpm -qa scylla\*')
```
Comment (roydahan) [inline upgrade_test.py]: remove

File: upgrade_test.py [hunk 4]
```diff
@@ -36,45 +37,57 @@ def upgrade_node(self, node):
...
         elif repo_file:
-            scylla_repo = get_data_path(repo_file)
-            node.remoter.send_files(scylla_repo, '/tmp/scylla.repo', verbose=True)
-            node.remoter.run('sudo cp /etc/yum.repos.d/scylla.repo ~/scylla.repo-backup')
-            node.remoter.run('sudo cp /tmp/scylla.repo /etc/yum.repos.d/scylla.repo')
+            if repo_file.startswith('http'):
+                node.remoter.run('sudo curl -L {} -o /etc/yum.repos.d/scylla.repo')
+            else:
```
Comment (roydahan) [inline upgrade_test.py]: we can remove this part entirely, we will work only with repo as http://...

File: upgrade_test.py [hunk 5]
```diff
@@ -36,45 +37,57 @@ def upgrade_node(self, node):
...
+                node.remoter.run('sudo yum remove scylla\* -y'.format(ver_suffix))
+                node.remoter.run('sudo yum install scylla{0} -y'.format(ver_suffix))
+        node.remoter.run('sudo systemctl start scylla-server.service')
         node.wait_db_up(verbose=True)
-        new_ver = node.remoter.run('rpm -qa scylla-server')
-        if orig_ver == new_ver:
-            self.log.error('scylla-server version isn\'t changed')
+        result = node.remoter.run('rpm -qa scylla\*server')
+        new_ver = result.stdout
+        assert orig_ver != new_ver, "scylla-server version isn't changed"
```
Comment (roydahan) [inline upgrade_test.py]: Please add running "nodetool status" command on this host.

File: upgrade_test.py [hunk 6]
```diff
@@ -36,45 +37,57 @@ def upgrade_node(self, node):
         repo_file = self.db_cluster.params.get('repo_file', None,  'scylla.repo.upgrade')
         new_version = self.db_cluster.params.get('new_version', None,  '')
         upgrade_node_packages = self.db_cluster.params.get('upgrade_node_packages')
+        smooth_upgrade_mode = self.db_cluster.params.get('smooth_upgrade_mode', default=True)
```
[the remark this answers is not in the document]
Comment (roydahan) [reply upgrade_test.py]: Please change the name to "remove_and_reinstall" or something meaningful.

File: upgrade_test.py [hunk 7]
```diff
@@ -115,15 +128,200 @@ def test_upgrade_cql_queries(self):
...
+        node.remoter.run('sudo chown root.root /etc/yum.repos.d/scylla.repo')
+        node.remoter.run('sudo chmod 644 /etc/yum.repos.d/scylla.repo')
+        node.remoter.run('sudo yum clean all')
+        if smooth_upgrade_mode:
+            # workaround Part 1
+            node.remoter.run('sudo yum remove scylla\*tools-core -y')
+            node.remoter.run('sudo yum downgrade scylla\* -y')
+        else:
+            node.remoter.run('sudo yum remove scylla\* -y')
+        # workaround Part 2
```
Comment (roydahan) [inline upgrade_test.py]: Is this part of the procedure we published?

File: upgrade_test.py [hunk 8]
```diff
@@ -115,15 +128,200 @@ def test_upgrade_cql_queries(self):
...
+
+    def test_partial_upgrade_rollback(self):
+        """
+        """
+
+        self.log.info('Starting c-s write workload for 5m')
+        stress_cmd = self.params.get('stress_cmd')
+        stress_queue = self.run_stress_thread(stress_cmd=stress_cmd)
+
+        self.log.info('Sleeping for 360s to let cassandra-stress populate some data before the mixed workload')
```
Comment (roydahan) [inline upgrade_test.py]: remove this part, not needed.

File: upgrade_test.py [hunk 9]
```diff
@@ -115,15 +128,200 @@ def test_upgrade_cql_queries(self):
...
+        result = node.remoter.run('rpm -qa scylla\*server')
+        new_ver = result.stdout
+        self.log.debug('original scylla-server version is %s, latest: %s' % (orig_ver, new_ver))
+        assert orig_ver != new_ver, "scylla-server version isn't changed"
+
+    def test_partial_upgrade_rollback(self):
+        """
+        """
+
+        self.log.info('Starting c-s write workload for 5m')
```
Comment (roydahan) [inline upgrade_test.py]: Not 5m, it should write all keys (n=1000xxxx)

File: upgrade_test.py [hunk 10]
```diff
@@ -115,15 +128,200 @@ def test_upgrade_cql_queries(self):
...
+        stress_cmd = self.params.get('stress_cmd')
+        stress_queue = self.run_stress_thread(stress_cmd=stress_cmd)
+
+        self.log.info('Sleeping for 360s to let cassandra-stress populate some data before the mixed workload')
+        time.sleep(360)
+
+        self.log.info('Starting c-s mixed workload for 60m')
```
Comment (roydahan) [inline upgrade_test.py]: remove mixed

File: upgrade_test.py [hunk 11]
```diff
@@ -115,15 +128,200 @@ def test_upgrade_cql_queries(self):
...
+            self.log.info('Upgrade Node %s ended', self.db_cluster.node_to_upgrade.name)
+
+        for i in indexes[:-1]:
+            self.log.info('Rollback Node %s begin', self.db_cluster.nodes[i].name)
+            self.rollback_node(self.db_cluster.nodes[i])
+            self.log.info('Rollback Node %s ended', self.db_cluster.nodes[i].name)
+
+        self.verify_stress_thread(stress_queue)
+
+    def test_upgrade_rollback_with_multi_workload(self):
```
Comment (roydahan) [inline upgrade_test.py]: I think we can unify all tests into one test scenario.
They aren't that different and it'll save time running one scenario.

we will need to run this every RC on all distributions.

File: upgrade_test.py [hunk 12]
```diff
@@ -115,15 +128,200 @@ def test_upgrade_cql_queries(self):
...
+            self.upgrade_node(self.db_cluster.node_to_upgrade)
+            self.log.info('Upgrade Node %s ended', self.db_cluster.node_to_upgrade.name)
+
+        stress_cmd_read_clall = self.params.get('stress_cmd_read_clall')
+        stress_queue_new = self.run_stress_thread(stress_cmd=stress_cmd_read_clall)
+
+        self.verify_stress_thread(stress_queue)
+        self.verify_stress_thread(stress_queue_new)
+
     def test_20_minutes(self):
```
Comment (roydahan) [inline upgrade_test.py]: let's remove old code.

File: tests/gce-upgrade.yaml [hunk 13]
```diff
@@ -0,0 +1,59 @@
+test_duration: 65
+
+# Single DC
+n_db_nodes: 5
+n_loaders: 1
+gce_datacenter: 'us-east1-b'
+
+stress_cmd: cassandra-stress write no-warmup cl=QUORUM duration=5m -schema 'replication(factor=2)' -port jmx=6868 -mode cql3 native -rate threads=1000 -errors ignore -pop seq=1..10000000 -log interval=5
+stress_cmd_1: cassandra-stress mixed no-warmup cl=QUORUM duration=60m -schema 'replication(factor=2)' -port jmx=6868 -mode cql3 native -rate threads=1000 -errors ignore -pop seq=1..10000000 -log interval=5
```
[the remark this answers is not in the document]
Comment (amoskong) [reply tests/gce-upgrade.yaml]: Then let's use `read` workload?

**Pull request closed without merging**